02-intermediate-ruby/references-and-values.md

```diff
@@ -0,0 +1,267 @@
+  after modification, param.object_id is 70144030241620
+  after reassignment, param.object_id is 70144030228060
+  with value ["new", "array"]
+  Finish reassign_parameter
+
+After reassign_parameter
+pets.object_id is 70144030241620
+with value ["dog", "parrot", "cat", "llama", "gecko"]
+```
+
+We can make a few interesting observations about this output:
+- The parameter inside the method has the same `object_id` as the variable we passed from outside
+- Modifications to the underlying object are visible outside the method
+- Reassigning the parameter with `=` does not reassign the outer variable
+
+This is exactly the same behavior we saw before, when we had two variables referencing the same object. From this we can conclude: **when you pass a variable as parameter, Ruby creates a new variable that references same object**.
+
+## Fixing the `short_strings` Method
+
+**Question:** Given what we've learned, how can we modify our `short_strings` method to do what we want?
+
+The answer is to modify the underlying object, rather than reassigning the parameter. Here's what the resulting code might look like:
+
+```ruby
+def short_strings_2(input)
+  input.each_with_index do |word, i|
+    # Slice characters 0 to 2
+    input[i] = word[0..2]
+  end
+end
+
+pets = ['dog', 'parrot', 'cat', 'llama']
+short_strings_2(pets)
+puts "#{pets}"
+```
+
+This produces the expected output. Note that we can't just say `word = word[0..2]`, for the same reason as above: that reassigns the block parameter `word` to a new string containing just the first 3 letters, but neither modifies nor reassigns the string in the array. Instead we reassign `input[i]`, which does what we want: change the value stored in the array.
+
+We could also use the `map!` enumerable method, since that modifies the original. `map` (without a `!`) would not work, because it creates a new array.
+
+## Other Objects
+
+We've talked a lot about arrays today, but this pattern holds true for all complex objects in Ruby: strings, hashes, instances of classes, etc. For example, consider the following code:
+
+```ruby
+# Reassign a string using +=
+def reassign_string(str)
+  str += ' reassigned'
+  puts "inside reassign_string, str is '#{str}'"
+end
+
+text = 'original'
+reassign_string(text)
+puts "outside reassign_string, text is '#{text}'"
+
+
+# Modify a string using the .concat() method
+def modify_string(str)
+  # str << ' modified' would do the same thing
+  str.concat(' modified')
+  puts "inside modify_string, str is '#{str}'"
+end
+
+text = 'original'
+modify_string(text)
+puts "outside modify_string, text is '#{text}'"
+```
+
+Primitive types like numbers, booleans and `nil` follow basically the same rules. The catch is there's no way to change the underlying value of a primitive without reassignment. In programming lingo, we say that these types are _immutable_. This means that whenever you change the value, Ruby makes a copy and changes that instead.
```

Comment:
As far as I'm aware, even though integers, floats, bools and nil are all "objects", the ruby interpreter represents them differently than "real" objects. The value is stored directly on the stack rather than in allocated memory. This means there's no "underlying object", and you can't have two references to the same integer.

That being said, "primitive" probably isn't the right word - see [this stackoverflow](https://stackoverflow.com/questions/18790442/are-there-primitive-types-in-ruby).